From 70c9755850b735724a9f8639c9ba959affa87004 Mon Sep 17 00:00:00 2001
From: Ryan Andri <ryanandri@linuxmail.org>
Date: Fri, 29 Dec 2017 04:06:23 +0700
Subject: [PATCH 02/16] input: add dt2w/s2w support.

Authored: Dennis Rassmann a.k.a showp1984

adapted by savoca at XDA-Developers.

original source: https://github.com/showp1984/bricked-hammerhead

Signed-off-by: Ryan Andri <ryanandri@linuxmail.org>
Change-Id: I47b72e812caba44d6d7453cff570e5b8101476fe
---
 drivers/input/touchscreen/Kconfig          |  17 +
 drivers/input/touchscreen/Makefile         |   4 +
 drivers/input/touchscreen/doubletap2wake.c | 401 +++++++++++++++++
 drivers/input/touchscreen/sweep2wake.c     | 484 +++++++++++++++++++++
 include/linux/input/doubletap2wake.h       |  28 ++
 include/linux/input/sweep2wake.h           |  28 ++
 6 files changed, 962 insertions(+)
 create mode 100644 drivers/input/touchscreen/doubletap2wake.c
 create mode 100644 drivers/input/touchscreen/sweep2wake.c
 create mode 100644 include/linux/input/doubletap2wake.h
 create mode 100644 include/linux/input/sweep2wake.h

diff --git a/drivers/input/touchscreen/Kconfig b/drivers/input/touchscreen/Kconfig
index 8aa1fb631..3cdfae3c9 100644
--- a/drivers/input/touchscreen/Kconfig
+++ b/drivers/input/touchscreen/Kconfig
@@ -1482,4 +1482,21 @@ source "drivers/input/touchscreen/td4310_e7/Kconfig"
 source "drivers/input/touchscreen/td4310_lansi_e7/Kconfig"
 source "drivers/input/touchscreen/vituralsar_e7/Kconfig"
 
+config TOUCHSCREEN_SWEEP2WAKE
+        tristate "Sweep2Wake for touchscreens"
+	select TOUCHSCREEN_PREVENT_SLEEP
+        default n
+
+config TOUCHSCREEN_DOUBLETAP2WAKE
+	tristate "DoubleTap2Wake for touchscreens"
+	select TOUCHSCREEN_PREVENT_SLEEP
+	default n
+
+config TOUCHSCREEN_PREVENT_SLEEP
+	bool "Inihibit sleep on modified touchscreen drivers"
+	default n
+	depends on TOUCHSCREEN_SWEEP2WAKE || TOUCHSCREEN_DOUBLETAP2WAKE
+	help
+	  This disables the sleep function of modified touchscreen drivers.
+
 endif
diff --git a/drivers/input/touchscreen/Makefile b/drivers/input/touchscreen/Makefile
index 15a8a97d3..2b7f52a12 100644
--- a/drivers/input/touchscreen/Makefile
+++ b/drivers/input/touchscreen/Makefile
@@ -132,3 +132,7 @@ obj-$(CONFIG_TOUCHSCREEN_NT36672_E7)	+= nt36672_e7/
 obj-$(CONFIG_TOUCHSCREEN_TD4310_E7)	+= td4310_e7/
 obj-$(CONFIG_TOUCHSCREEN_TD4310_LANSI_E7)	+= td4310_lansi_e7/
 obj-$(CONFIG_TOUCHSCREEN_VITURALSAR_E7)	+= vituralsar_e7/
+
+# dt2w/s2w
+obj-$(CONFIG_TOUCHSCREEN_SWEEP2WAKE)	+= sweep2wake.o
+obj-$(CONFIG_TOUCHSCREEN_DOUBLETAP2WAKE)	+= doubletap2wake.o
diff --git a/drivers/input/touchscreen/doubletap2wake.c b/drivers/input/touchscreen/doubletap2wake.c
new file mode 100644
index 000000000..b9d0cee29
--- /dev/null
+++ b/drivers/input/touchscreen/doubletap2wake.c
@@ -0,0 +1,401 @@
+/*
+ * drivers/input/touchscreen/doubletap2wake.c
+ *
+ *
+ * Copyright (c) 2013, Dennis Rassmann <showp1984@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/types.h>
+#include <linux/delay.h>
+#include <linux/init.h>
+#include <linux/err.h>
+#include <linux/input/doubletap2wake.h>
+#include <linux/slab.h>
+#include <linux/workqueue.h>
+#include <linux/input.h>
+#include <linux/hrtimer.h>
+#include <asm-generic/cputime.h>
+
+/* uncomment since no touchscreen defines android touch, do that here */
+//#define ANDROID_TOUCH_DECLARED
+
+/* if Sweep2Wake is compiled it will already have taken care of this */
+#ifdef CONFIG_TOUCHSCREEN_SWEEP2WAKE
+#define ANDROID_TOUCH_DECLARED
+#endif
+
+/* Version, author, desc, etc */
+#define DRIVER_AUTHOR "Dennis Rassmann <showp1984@gmail.com>"
+#define DRIVER_DESCRIPTION "Doubletap2wake for almost any device"
+#define DRIVER_VERSION "1.0"
+#define LOGTAG "[doubletap2wake]: "
+
+MODULE_AUTHOR(DRIVER_AUTHOR);
+MODULE_DESCRIPTION(DRIVER_DESCRIPTION);
+MODULE_VERSION(DRIVER_VERSION);
+MODULE_LICENSE("GPLv2");
+
+/* Tuneables */
+#define DT2W_DEBUG		0
+#define DT2W_DEFAULT		0
+
+#define DT2W_PWRKEY_DUR		60
+#define DT2W_FEATHER		200
+#define DT2W_TIME		700
+
+/* Resources */
+int dt2w_switch = DT2W_DEFAULT;
+bool dt2w_scr_suspended = false;
+static cputime64_t tap_time_pre = 0;
+static int touch_x = 0, touch_y = 0, touch_nr = 0, x_pre = 0, y_pre = 0;
+static bool touch_x_called = false, touch_y_called = false, touch_cnt = true;
+static bool exec_count = true;
+static struct input_dev * doubletap2wake_pwrdev;
+static DEFINE_MUTEX(pwrkeyworklock);
+static struct workqueue_struct *dt2w_input_wq;
+static struct work_struct dt2w_input_work;
+
+/* Read cmdline for dt2w */
+static int __init read_dt2w_cmdline(char *dt2w)
+{
+	if (strcmp(dt2w, "1") == 0) {
+		pr_info("[cmdline_dt2w]: DoubleTap2Wake enabled. | dt2w='%s'\n", dt2w);
+		dt2w_switch = 1;
+	} else if (strcmp(dt2w, "0") == 0) {
+		pr_info("[cmdline_dt2w]: DoubleTap2Wake disabled. | dt2w='%s'\n", dt2w);
+		dt2w_switch = 0;
+	} else {
+		pr_info("[cmdline_dt2w]: No valid input found. Going with default: | dt2w='%u'\n", dt2w_switch);
+	}
+	return 1;
+}
+__setup("dt2w=", read_dt2w_cmdline);
+
+/* reset on finger release */
+static void doubletap2wake_reset(void) {
+	exec_count = true;
+	touch_nr = 0;
+	tap_time_pre = 0;
+	x_pre = 0;
+	y_pre = 0;
+	touch_cnt = false;
+}
+
+/* PowerKey work func */
+static void doubletap2wake_presspwr(struct work_struct * doubletap2wake_presspwr_work) {
+	if (!mutex_trylock(&pwrkeyworklock))
+                return;
+	input_event(doubletap2wake_pwrdev, EV_KEY, KEY_POWER, 1);
+	input_event(doubletap2wake_pwrdev, EV_SYN, 0, 0);
+	msleep(DT2W_PWRKEY_DUR);
+	input_event(doubletap2wake_pwrdev, EV_KEY, KEY_POWER, 0);
+	input_event(doubletap2wake_pwrdev, EV_SYN, 0, 0);
+	msleep(DT2W_PWRKEY_DUR);
+        mutex_unlock(&pwrkeyworklock);
+	return;
+}
+static DECLARE_WORK(doubletap2wake_presspwr_work, doubletap2wake_presspwr);
+
+/* PowerKey trigger */
+static void doubletap2wake_pwrtrigger(void) {
+	schedule_work(&doubletap2wake_presspwr_work);
+        return;
+}
+
+/* unsigned */
+static unsigned int calc_feather(int coord, int prev_coord) {
+	int calc_coord = 0;
+	calc_coord = coord-prev_coord;
+	if (calc_coord < 0)
+		calc_coord = calc_coord * (-1);
+	return calc_coord;
+}
+
+/* init a new touch */
+static void new_touch(int x, int y) {
+	tap_time_pre = ktime_to_ms(ktime_get());
+	x_pre = x;
+	y_pre = y;
+	touch_nr++;
+}
+
+/* Doubletap2wake main function */
+static void detect_doubletap2wake(int x, int y, bool st)
+{
+        bool single_touch = st;
+#if DT2W_DEBUG
+        pr_info(LOGTAG"x,y(%4d,%4d) single:%s\n",
+                x, y, (single_touch) ? "true" : "false");
+#endif
+	if ((single_touch) && (dt2w_switch > 0) && (exec_count) && (touch_cnt)) {
+		
+		if ((ktime_to_ms(ktime_get())-tap_time_pre) >= DT2W_TIME)
+			doubletap2wake_reset();
+		
+		if (touch_nr == 0) {
+			new_touch(x, y);
+		} else if (touch_nr == 1) {
+			if ((calc_feather(x, x_pre) < DT2W_FEATHER) &&
+			    (calc_feather(y, y_pre) < DT2W_FEATHER)) {
+				pr_info(LOGTAG"ON\n");
+				exec_count = false;
+				doubletap2wake_pwrtrigger();
+				doubletap2wake_reset();
+			} else {
+				doubletap2wake_reset();
+				new_touch(x, y);
+			}
+		}
+		/*if ((touch_nr > 1)) {
+			pr_info(LOGTAG"ON\n");
+			exec_count = false;
+			doubletap2wake_pwrtrigger();
+			doubletap2wake_reset();
+		}*/
+	}
+}
+
+static void dt2w_input_callback(struct work_struct *unused) {
+
+	detect_doubletap2wake(touch_x, touch_y, true);
+
+	return;
+}
+
+static void dt2w_input_event(struct input_handle *handle, unsigned int type,
+				unsigned int code, int value) {
+#if DT2W_DEBUG
+	pr_info("doubletap2wake: code: %s|%u, val: %i\n",
+		((code==ABS_MT_POSITION_X) ? "X" :
+		(code==ABS_MT_POSITION_Y) ? "Y" :
+		(code==ABS_MT_TRACKING_ID) ? "ID" :
+		"undef"), code, value);
+#endif
+	if (!dt2w_scr_suspended)
+		return;
+
+	if (code == ABS_MT_SLOT) {
+		doubletap2wake_reset();
+		return;
+	}
+
+	if (code == ABS_MT_TRACKING_ID && value == -1) {
+		touch_cnt = true;
+		return;
+	}
+
+	if (code == ABS_MT_POSITION_X) {
+		touch_x = value;
+		touch_x_called = true;
+	}
+
+	if (code == ABS_MT_POSITION_Y) {
+		touch_y = value;
+		touch_y_called = true;
+	}
+
+	if ((touch_x_called || touch_y_called) && touch_cnt)  {
+		touch_x_called = false;
+		touch_y_called = false;
+		queue_work_on(0, dt2w_input_wq, &dt2w_input_work);
+	}
+}
+
+static int input_dev_filter(struct input_dev *dev) {
+	if (strstr(dev->name, "touch") ||
+		strstr(dev->name, "synaptics_dsx_i2c")) {
+		return 0;
+	} else {
+		return 1;
+	}
+}
+
+static int dt2w_input_connect(struct input_handler *handler,
+				struct input_dev *dev, const struct input_device_id *id) {
+	struct input_handle *handle;
+	int error;
+
+	if (input_dev_filter(dev))
+		return -ENODEV;
+
+	handle = kzalloc(sizeof(struct input_handle), GFP_KERNEL);
+	if (!handle)
+		return -ENOMEM;
+
+	handle->dev = dev;
+	handle->handler = handler;
+	handle->name = "dt2w";
+
+	error = input_register_handle(handle);
+	if (error)
+		goto err2;
+
+	error = input_open_device(handle);
+	if (error)
+		goto err1;
+
+	return 0;
+err1:
+	input_unregister_handle(handle);
+err2:
+	kfree(handle);
+	return error;
+}
+
+static void dt2w_input_disconnect(struct input_handle *handle) {
+	input_close_device(handle);
+	input_unregister_handle(handle);
+	kfree(handle);
+}
+
+static const struct input_device_id dt2w_ids[] = {
+	{ .driver_info = 1 },
+	{ },
+};
+
+static struct input_handler dt2w_input_handler = {
+	.event		= dt2w_input_event,
+	.connect	= dt2w_input_connect,
+	.disconnect	= dt2w_input_disconnect,
+	.name		= "dt2w_inputreq",
+	.id_table	= dt2w_ids,
+};
+
+/*
+ * SYSFS stuff below here
+ */
+static ssize_t dt2w_doubletap2wake_show(struct device *dev,
+		struct device_attribute *attr, char *buf)
+{
+	size_t count = 0;
+
+	count += sprintf(buf, "%d\n", dt2w_switch);
+
+	return count;
+}
+
+static ssize_t dt2w_doubletap2wake_dump(struct device *dev,
+		struct device_attribute *attr, const char *buf, size_t count)
+{
+	if (buf[0] >= '0' && buf[0] <= '2' && buf[1] == '\n')
+                if (dt2w_switch != buf[0] - '0')
+		        dt2w_switch = buf[0] - '0';
+
+	return count;
+}
+
+static DEVICE_ATTR(doubletap2wake, (S_IWUSR|S_IRUGO),
+	dt2w_doubletap2wake_show, dt2w_doubletap2wake_dump);
+
+static ssize_t dt2w_version_show(struct device *dev,
+		struct device_attribute *attr, char *buf)
+{
+	size_t count = 0;
+
+	count += sprintf(buf, "%s\n", DRIVER_VERSION);
+
+	return count;
+}
+
+static ssize_t dt2w_version_dump(struct device *dev,
+		struct device_attribute *attr, const char *buf, size_t count)
+{
+	return count;
+}
+
+static DEVICE_ATTR(doubletap2wake_version, (S_IWUSR|S_IRUGO),
+	dt2w_version_show, dt2w_version_dump);
+
+/*
+ * INIT / EXIT stuff below here
+ */
+#ifdef ANDROID_TOUCH_DECLARED
+extern struct kobject *android_touch_kobj;
+#else
+struct kobject *android_touch_kobj;
+EXPORT_SYMBOL_GPL(android_touch_kobj);
+#endif
+static int __init doubletap2wake_init(void)
+{
+	int rc = 0;
+
+	doubletap2wake_pwrdev = input_allocate_device();
+	if (!doubletap2wake_pwrdev) {
+		pr_err("Can't allocate suspend autotest power button\n");
+		goto err_alloc_dev;
+	}
+
+	input_set_capability(doubletap2wake_pwrdev, EV_KEY, KEY_POWER);
+	doubletap2wake_pwrdev->name = "dt2w_pwrkey";
+	doubletap2wake_pwrdev->phys = "dt2w_pwrkey/input0";
+
+	rc = input_register_device(doubletap2wake_pwrdev);
+	if (rc) {
+		pr_err("%s: input_register_device err=%d\n", __func__, rc);
+		goto err_input_dev;
+	}
+
+	dt2w_input_wq = create_workqueue("dt2wiwq");
+	if (!dt2w_input_wq) {
+		pr_err("%s: Failed to create dt2wiwq workqueue\n", __func__);
+		return -EFAULT;
+	}
+	INIT_WORK(&dt2w_input_work, dt2w_input_callback);
+	rc = input_register_handler(&dt2w_input_handler);
+	if (rc)
+		pr_err("%s: Failed to register dt2w_input_handler\n", __func__);
+
+#ifndef ANDROID_TOUCH_DECLARED
+	android_touch_kobj = kobject_create_and_add("android_touch", NULL) ;
+	if (android_touch_kobj == NULL) {
+		pr_warn("%s: android_touch_kobj create_and_add failed\n", __func__);
+	}
+#endif
+	rc = sysfs_create_file(android_touch_kobj, &dev_attr_doubletap2wake.attr);
+	if (rc) {
+		pr_warn("%s: sysfs_create_file failed for doubletap2wake\n", __func__);
+	}
+
+	rc = sysfs_create_file(android_touch_kobj, &dev_attr_doubletap2wake_version.attr);
+	if (rc) {
+		pr_warn("%s: sysfs_create_file failed for doubletap2wake_version\n", __func__);
+	}
+
+err_input_dev:
+	input_free_device(doubletap2wake_pwrdev);
+err_alloc_dev:
+	pr_info(LOGTAG"%s done\n", __func__);
+
+	return 0;
+}
+
+static void __exit doubletap2wake_exit(void)
+{
+#ifndef ANDROID_TOUCH_DECLARED
+	kobject_del(android_touch_kobj);
+#endif
+	input_unregister_handler(&dt2w_input_handler);
+	destroy_workqueue(dt2w_input_wq);
+	input_unregister_device(doubletap2wake_pwrdev);
+	input_free_device(doubletap2wake_pwrdev);
+	return;
+}
+
+late_initcall(doubletap2wake_init);
+module_exit(doubletap2wake_exit);
diff --git a/drivers/input/touchscreen/sweep2wake.c b/drivers/input/touchscreen/sweep2wake.c
new file mode 100644
index 000000000..aaed4fc68
--- /dev/null
+++ b/drivers/input/touchscreen/sweep2wake.c
@@ -0,0 +1,484 @@
+/*
+ * drivers/input/touchscreen/sweep2wake.c
+ *
+ *
+ * Copyright (c) 2013, Dennis Rassmann <showp1984@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/types.h>
+#include <linux/delay.h>
+#include <linux/init.h>
+#include <linux/err.h>
+#include <linux/input/sweep2wake.h>
+#include <linux/slab.h>
+#include <linux/workqueue.h>
+#include <linux/input.h>
+#include <linux/hrtimer.h>
+
+/* uncomment since no touchscreen defines android touch, do that here */
+//#define ANDROID_TOUCH_DECLARED
+
+/* Version, author, desc, etc */
+#define DRIVER_AUTHOR "Dennis Rassmann <showp1984@gmail.com>"
+#define DRIVER_DESCRIPTION "Sweep2wake for almost any device"
+#define DRIVER_VERSION "1.5"
+#define LOGTAG "[sweep2wake]: "
+
+MODULE_AUTHOR(DRIVER_AUTHOR);
+MODULE_DESCRIPTION(DRIVER_DESCRIPTION);
+MODULE_VERSION(DRIVER_VERSION);
+MODULE_LICENSE("GPLv2");
+
+#ifdef CONFIG_MACH_MSM8974_HAMMERHEAD
+/* Hammerhead aka Nexus 5 */
+#define S2W_Y_MAX               1920
+#define S2W_X_MAX               1080
+#define S2W_Y_LIMIT             S2W_Y_MAX-130
+#define S2W_X_B1                400
+#define S2W_X_B2                700
+#define S2W_X_FINAL             250
+#elif defined(CONFIG_MACH_APQ8064_MAKO)
+/* Mako aka Nexus 4 */
+#define S2W_Y_LIMIT             2350
+#define S2W_X_MAX               1540
+#define S2W_X_B1                500
+#define S2W_X_B2                1000
+#define S2W_X_FINAL             300
+#elif defined(CONFIG_MACH_APQ8064_FLO)
+/* Flo/Deb aka Nexus 7 2013 */
+#define S2W_Y_MAX               2240
+#define S2W_X_MAX               1344
+#define S2W_Y_LIMIT             S2W_Y_MAX-110
+#define S2W_X_B1                500
+#define S2W_X_B2                700
+#define S2W_X_FINAL             450
+#else
+/* defaults */
+#define S2W_Y_MAX				1280
+#define S2W_X_MAX				720
+#define S2W_Y_LIMIT				S2W_Y_MAX-70
+#define S2W_X_B1				200
+#define S2W_X_B2				400
+#define S2W_X_FINAL				180
+#endif
+
+/* Resources */
+int s2w_switch = 0;
+bool s2w_scr_suspended = false;
+static int s2w_debug = 0;
+static int s2w_pwrkey_dur = 60;
+static int touch_x = 0, touch_y = 0;
+static bool touch_x_called = false, touch_y_called = false;
+static bool exec_count = true;
+static bool scr_on_touch = false, barrier[2] = {false, false};
+static struct input_dev * sweep2wake_pwrdev;
+static DEFINE_MUTEX(pwrkeyworklock);
+static struct workqueue_struct *s2w_input_wq;
+static struct work_struct s2w_input_work;
+
+/* Read cmdline for s2w */
+static int __init read_s2w_cmdline(char *s2w)
+{
+	if (strcmp(s2w, "1") == 0) {
+		pr_info("[cmdline_s2w]: Sweep2Wake enabled. | s2w='%s'\n", s2w);
+		s2w_switch = 1;
+	} else if (strcmp(s2w, "2") == 0) {
+		pr_info("[cmdline_s2w]: Sweep2Sleep enabled. | s2w='%s'\n", s2w);
+		s2w_switch = 2;
+	} else if (strcmp(s2w, "0") == 0) {
+		pr_info("[cmdline_s2w]: Sweep2Wake disabled. | s2w='%s'\n", s2w);
+		s2w_switch = 0;
+	} else {
+		pr_info("[cmdline_s2w]: No valid input found. Going with default: | s2w='%u'\n", s2w_switch);
+	}
+	return 1;
+}
+__setup("s2w=", read_s2w_cmdline);
+
+/* PowerKey work func */
+static void sweep2wake_presspwr(struct work_struct * sweep2wake_presspwr_work) {
+	if (!mutex_trylock(&pwrkeyworklock))
+                return;
+	input_event(sweep2wake_pwrdev, EV_KEY, KEY_POWER, 1);
+	input_event(sweep2wake_pwrdev, EV_SYN, 0, 0);
+	msleep(s2w_pwrkey_dur);
+	input_event(sweep2wake_pwrdev, EV_KEY, KEY_POWER, 0);
+	input_event(sweep2wake_pwrdev, EV_SYN, 0, 0);
+	msleep(s2w_pwrkey_dur);
+        mutex_unlock(&pwrkeyworklock);
+	return;
+}
+static DECLARE_WORK(sweep2wake_presspwr_work, sweep2wake_presspwr);
+
+/* PowerKey trigger */
+static void sweep2wake_pwrtrigger(void) {
+	schedule_work(&sweep2wake_presspwr_work);
+        return;
+}
+
+/* reset on finger release */
+static void sweep2wake_reset(void) {
+	exec_count = true;
+	barrier[0] = false;
+	barrier[1] = false;
+	scr_on_touch = false;
+}
+
+/* Sweep2wake main function */
+static void detect_sweep2wake(int x, int y)
+{
+	int prevx = 0, nextx = 0;
+
+	if (s2w_debug)
+		pr_info(LOGTAG"x: %d, y: %d\n", x, y);
+
+	//left->right
+	if (s2w_scr_suspended == true) {
+		prevx = 0;
+		nextx = S2W_X_B1;
+		if ((barrier[0] == true) ||
+		   ((x > prevx) &&
+		    (x < nextx) &&
+		    (y > 0))) {
+			prevx = nextx;
+			nextx = S2W_X_B2;
+			barrier[0] = true;
+			if ((barrier[1] == true) ||
+			   ((x > prevx) &&
+			    (x < nextx) &&
+			    (y > 0))) {
+				prevx = nextx;
+				barrier[1] = true;
+				if ((x > prevx) &&
+				    (y > 0)) {
+					if (x > (S2W_X_MAX - S2W_X_FINAL)) {
+						if (exec_count) {
+							pr_info(LOGTAG"ON\n");
+							sweep2wake_pwrtrigger();
+							exec_count = false;
+						}
+					}
+				}
+			}
+		}
+	//right->left
+	} else if ((s2w_scr_suspended == false) && (s2w_switch > 0)) {
+		scr_on_touch=true;
+		prevx = (S2W_X_MAX - S2W_X_FINAL);
+		nextx = S2W_X_B2;
+		if ((barrier[0] == true) ||
+		   ((x < prevx) &&
+		    (x > nextx) &&
+		    (y > S2W_Y_LIMIT))) {
+			prevx = nextx;
+			nextx = S2W_X_B1;
+			barrier[0] = true;
+			if ((barrier[1] == true) ||
+			   ((x < prevx) &&
+			    (x > nextx) &&
+			    (y > S2W_Y_LIMIT))) {
+				prevx = nextx;
+				barrier[1] = true;
+				if ((x < prevx) &&
+				    (y > S2W_Y_LIMIT)) {
+					if (x < S2W_X_FINAL) {
+						if (exec_count) {
+							pr_info(LOGTAG"OFF\n");
+							sweep2wake_pwrtrigger();
+							exec_count = false;
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+static void s2w_input_callback(struct work_struct *unused) {
+
+	if (s2w_switch)
+		detect_sweep2wake(touch_x, touch_y);
+
+	return;
+}
+
+static void s2w_input_event(struct input_handle *handle, unsigned int type,
+				unsigned int code, int value)
+{
+	if ((!s2w_switch) || ((s2w_scr_suspended) && (s2w_switch > 1)))
+		return;
+
+	if (code == ABS_MT_SLOT) {
+		sweep2wake_reset();
+		return;
+	}
+
+	if (code == ABS_MT_TRACKING_ID && value == -1) {
+		sweep2wake_reset();
+		return;
+	}
+
+	if (code == ABS_MT_POSITION_X) {
+		touch_x = value;
+		touch_x_called = true;
+	}
+
+	if (code == ABS_MT_POSITION_Y) {
+		touch_y = value;
+		touch_y_called = true;
+	}
+
+	if (touch_x_called && touch_y_called) {
+		touch_x_called = false;
+		touch_y_called = false;
+		queue_work_on(0, s2w_input_wq, &s2w_input_work);
+	}
+}
+
+static int input_dev_filter(struct input_dev *dev) {
+	if (strstr(dev->name, "touch") ||
+		strstr(dev->name, "synaptics_dsx_i2c")) {
+		return 0;
+	} else {
+		return 1;
+	}
+}
+
+static int s2w_input_connect(struct input_handler *handler,
+				struct input_dev *dev, const struct input_device_id *id) {
+	struct input_handle *handle;
+	int error;
+
+	if (input_dev_filter(dev))
+		return -ENODEV;
+
+	handle = kzalloc(sizeof(struct input_handle), GFP_KERNEL);
+	if (!handle)
+		return -ENOMEM;
+
+	handle->dev = dev;
+	handle->handler = handler;
+	handle->name = "s2w";
+
+	error = input_register_handle(handle);
+	if (error)
+		goto err2;
+
+	error = input_open_device(handle);
+	if (error)
+		goto err1;
+
+	return 0;
+err1:
+	input_unregister_handle(handle);
+err2:
+	kfree(handle);
+	return error;
+}
+
+static void s2w_input_disconnect(struct input_handle *handle) {
+	input_close_device(handle);
+	input_unregister_handle(handle);
+	kfree(handle);
+}
+
+static const struct input_device_id s2w_ids[] = {
+	{ .driver_info = 1 },
+	{ },
+};
+
+static struct input_handler s2w_input_handler = {
+	.event		= s2w_input_event,
+	.connect	= s2w_input_connect,
+	.disconnect	= s2w_input_disconnect,
+	.name		= "s2w_inputreq",
+	.id_table	= s2w_ids,
+};
+
+/*
+ * SYSFS stuff below here
+ */
+static ssize_t s2w_sweep2wake_show(struct device *dev,
+		struct device_attribute *attr, char *buf)
+{
+	size_t count = 0;
+
+	count += sprintf(buf, "%d\n", s2w_switch);
+
+	return count;
+}
+
+static ssize_t s2w_sweep2wake_dump(struct device *dev,
+		struct device_attribute *attr, const char *buf, size_t count)
+{
+	if (buf[0] >= '0' && buf[0] <= '2' && buf[1] == '\n')
+                if (s2w_switch != buf[0] - '0')
+		        s2w_switch = buf[0] - '0';
+
+	return count;
+}
+
+static DEVICE_ATTR(sweep2wake, (S_IWUSR|S_IRUGO),
+	s2w_sweep2wake_show, s2w_sweep2wake_dump);
+
+static ssize_t s2w_debug_show(struct device *dev,
+		struct device_attribute *attr, char *buf)
+{
+	return sprintf(buf, "%d\n", s2w_debug);
+}
+
+static ssize_t s2w_debug_dump(struct device *dev,
+		struct device_attribute *attr, const char *buf, size_t count)
+{
+	if (buf[0] >= '0' && buf[0] <= '1' && buf[1] == '\n')
+		if (s2w_debug != buf[0] - '0')
+			s2w_debug = buf[0] - '0';
+
+	return count;
+}
+
+static DEVICE_ATTR(sweep2wake_debug, (S_IWUSR|S_IRUGO),
+	s2w_debug_show, s2w_debug_dump);
+
+static ssize_t s2w_pwrkey_dur_show(struct device *dev,
+		struct device_attribute *attr, char *buf)
+{
+	return sprintf(buf, "%d\n", s2w_pwrkey_dur);
+}
+
+static ssize_t s2w_pwrkey_dur_dump(struct device *dev,
+		struct device_attribute *attr, const char *buf, size_t count)
+{
+	int val, err;
+
+	err = kstrtoint(buf, 0, &val);
+	if (err < 0) {
+		printk(KERN_ERR "%s: Failed to convert value.\n", __func__);
+		return -EINVAL;
+	}
+
+	if ((!val) || (val > 3000)) {
+		printk(KERN_ERR "%s: value is out of range.\n", __func__);
+		return -EINVAL;
+	} else
+		s2w_pwrkey_dur = val;
+
+	return count;
+}
+
+static DEVICE_ATTR(sweep2wake_pwrkey_dur, (S_IWUSR|S_IRUGO),
+	s2w_pwrkey_dur_show, s2w_pwrkey_dur_dump);
+
+static ssize_t s2w_version_show(struct device *dev,
+		struct device_attribute *attr, char *buf)
+{
+	size_t count = 0;
+
+	count += sprintf(buf, "%s\n", DRIVER_VERSION);
+
+	return count;
+}
+
+static DEVICE_ATTR(sweep2wake_version, (S_IWUSR|S_IRUGO),
+	s2w_version_show, NULL);
+
+/*
+ * INIT / EXIT stuff below here
+ */
+#ifdef ANDROID_TOUCH_DECLARED
+extern struct kobject *android_touch_kobj;
+#else
+struct kobject *android_touch_kobj;
+EXPORT_SYMBOL_GPL(android_touch_kobj);
+#endif
+static int __init sweep2wake_init(void)
+{
+	int rc = 0;
+
+	sweep2wake_pwrdev = input_allocate_device();
+	if (!sweep2wake_pwrdev) {
+		pr_err("Can't allocate suspend autotest power button\n");
+		goto err_alloc_dev;
+	}
+
+	input_set_capability(sweep2wake_pwrdev, EV_KEY, KEY_POWER);
+	sweep2wake_pwrdev->name = "s2w_pwrkey";
+	sweep2wake_pwrdev->phys = "s2w_pwrkey/input0";
+
+	rc = input_register_device(sweep2wake_pwrdev);
+	if (rc) {
+		pr_err("%s: input_register_device err=%d\n", __func__, rc);
+		goto err_input_dev;
+	}
+
+	s2w_input_wq = create_workqueue("s2wiwq");
+	if (!s2w_input_wq) {
+		pr_err("%s: Failed to create s2wiwq workqueue\n", __func__);
+		return -EFAULT;
+	}
+	INIT_WORK(&s2w_input_work, s2w_input_callback);
+	rc = input_register_handler(&s2w_input_handler);
+	if (rc)
+		pr_err("%s: Failed to register s2w_input_handler\n", __func__);
+
+#ifndef ANDROID_TOUCH_DECLARED
+	android_touch_kobj = kobject_create_and_add("android_touch", NULL) ;
+	if (android_touch_kobj == NULL) {
+		pr_warn("%s: android_touch_kobj create_and_add failed\n", __func__);
+	}
+#endif
+	rc = sysfs_create_file(android_touch_kobj, &dev_attr_sweep2wake.attr);
+	if (rc) {
+		pr_warn("%s: sysfs_create_file failed for sweep2wake\n", __func__);
+	}
+	rc = sysfs_create_file(android_touch_kobj, &dev_attr_sweep2wake_debug.attr);
+	if (rc) {
+		pr_warn("%s: sysfs_create_file failed for sweep2wake_debug\n", __func__);
+	}
+	rc = sysfs_create_file(android_touch_kobj, &dev_attr_sweep2wake_pwrkey_dur.attr);
+	if (rc) {
+		pr_warn("%s: sysfs_create_file failed for sweep2wake_pwrkey_dur\n", __func__);
+	}
+	rc = sysfs_create_file(android_touch_kobj, &dev_attr_sweep2wake_version.attr);
+	if (rc) {
+		pr_warn("%s: sysfs_create_file failed for sweep2wake_version\n", __func__);
+	}
+
+err_input_dev:
+	input_free_device(sweep2wake_pwrdev);
+err_alloc_dev:
+	pr_info(LOGTAG"%s done\n", __func__);
+
+	return 0;
+}
+
+static void __exit sweep2wake_exit(void)
+{
+#ifndef ANDROID_TOUCH_DECLARED
+	kobject_del(android_touch_kobj);
+#endif
+	input_unregister_handler(&s2w_input_handler);
+	destroy_workqueue(s2w_input_wq);
+	input_unregister_device(sweep2wake_pwrdev);
+	input_free_device(sweep2wake_pwrdev);
+	return;
+}
+
+late_initcall(sweep2wake_init);
+module_exit(sweep2wake_exit);
diff --git a/include/linux/input/doubletap2wake.h b/include/linux/input/doubletap2wake.h
new file mode 100644
index 000000000..0f2305e6c
--- /dev/null
+++ b/include/linux/input/doubletap2wake.h
@@ -0,0 +1,28 @@
+/*
+ * include/linux/input/doubletap2wake.h
+ *
+ * Copyright (c) 2013, Dennis Rassmann <showp1984@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#ifndef _LINUX_DOUBLETAP2WAKE_H
+#define _LINUX_DOUBLETAP2WAKE_H
+
+extern int dt2w_switch;
+
+extern bool dt2w_scr_suspended;
+
+#endif	/* _LINUX_DOUBLETAP2WAKE_H */
diff --git a/include/linux/input/sweep2wake.h b/include/linux/input/sweep2wake.h
new file mode 100644
index 000000000..2f75047bc
--- /dev/null
+++ b/include/linux/input/sweep2wake.h
@@ -0,0 +1,28 @@
+/*
+ * include/linux/input/sweep2wake.h
+ *
+ * Copyright (c) 2013, Dennis Rassmann <showp1984@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#ifndef _LINUX_SWEEP2WAKE_H
+#define _LINUX_SWEEP2WAKE_H
+
+extern int s2w_switch;
+
+extern bool s2w_scr_suspended;
+
+#endif	/* _LINUX_SWEEP2WAKE_H */
-- 
2.55.0

